### ISSUE TITLE
Generic arguments do not match, i.e Type "TSource" cannot be assigned to type "TSource"

### ISSUE THREAD

**Issue by dbrattli** [microsoft/pyright#1235]
## Environment data

* Language Server version: Pylance language server 2020.12.1-pre.1 (pyright ea581b3f)
* OS and version: Mac OS X 11.0.1
* Python version (& distribution if applicable, e.g. Anaconda): Python 3.9 (brew)

## Expected behaviour

Generic arguments should match up. This used to work but got broken in 2020.12.1-pre.1 release after microsoft/pylance-release#636 got fixed.

## Actual behaviour

![Screenshot 2020-12-03 at 11 01 58](https://user-images.githubusercontent.com/849479/100994616-107d3500-3557-11eb-91ce-e15499429fc3.png)
![Screenshot 2020-12-03 at 11 02 07](https://user-images.githubusercontent.com/849479/100994624-11ae6200-3557-11eb-80f1-4f33e66e36ed.png)
![Screenshot 2020-12-03 at 11 02 13](https://user-images.githubusercontent.com/849479/100994634-13782580-3557-11eb-92e0-d640a133c14e.png)

## Logs

```
[Info  - 7:43:16 AM] Pylance language server 2020.12.1-pre.1 (pyright ea581b3f) starting
[Info  - 7:43:16 AM] Server root directory: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.12.1-pre.1/dist
[Error - 7:43:16 AM] stubPath /Users/dbrattli/Developer/Github/Expression/typings is not a valid directory.
[Info  - 9:27:27 AM] No configuration file found.
[Info  - 9:27:27 AM] Setting pythonPath for service "Expression": "/usr/local/opt/python@3.9/bin/python3"
...
```

## Code Snippet / Additional information

```python
from typing import Callable, Iterable, Optional, Tuple, TypeVar

TSource = TypeVar("TSource")
TState = TypeVar("TState")


class Seq(Iterable[TSource]):
    @staticmethod
    def unfold(generator: Callable[[TState], Optional[Tuple[TSource, TState]]], state: TState) -> Iterable[TSource]:
        while True:
            result = generator(state)
            if result is None:
                break

            item, state = result
            yield item


def unfold(generator: Callable[[TState], Optional[Tuple[TSource, TState]]]) -> Callable[[TState], Iterable[TSource]]:
    def _unfold(state: TState) -> Iterable[TSource]:
        return Seq.unfold(generator, state)

    return _unfold
```

**Comment by erictraut** [comment]
Thanks for the bug report. This is related to #1217.

**Comment by dbrattli** [comment]
As a note to this I just discovered that reversing the dependency fixes the typing problem, ie. the staticmethod depends on the unbound function and not the other way around. 

```py
from typing import Callable, Iterable, Optional, Tuple, TypeVar

TSource = TypeVar("TSource")
TState = TypeVar("TState")


class Seq(Iterable[TSource]):
    @staticmethod
    def unfold(generator: Callable[[TState], Optional[Tuple[TSource, TState]]], state: TState) -> Iterable[TSource]:
        return unfold(generator)(state)


def unfold(generator: Callable[[TState], Optional[Tuple[TSource, TState]]]) -> Callable[[TState], Iterable[TSource]]:
    def _unfold(state: TState) -> Iterable[TSource]:
        while True:
            result = generator(state)
            if result is None:
                break

            item, state = result
            yield item

    return _unfold
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by dbrattli** [comment]
Thanks for fixing! Looking forward to testing this once it's ready.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.93, which I just published. It will also be included in the next release of Pylance.
